File: 2018/javascript/day_01.js
function main() {
  let input = "";
  process.stdin.on("data", data => input += data.toString().trim())
  process.stdin.on("end", () => {
    console.log(star_01(input))
    console.log(star_02(input))
  })
}

function star_01(input) {
  let freq = 0
  for (const line of input.split('\n')) {
    freq += Number.parseInt(line)
  }
  return freq
}

function star_02(input) {
  let freq = 0
  const seen = new Set([freq])
  const inputs = input.split('\n').map(str => Number.parseInt(str))
  while (true) {
    for (const line of inputs) {
      freq += Number.parseInt(line)
      if (seen.has(freq)) {
        return freq
      }
      seen.add(freq)
    }
  }
}

main()

File: 2018/elixir/day_01.exs
defmodule AdventOfCode.Day_01 do
  def star_01(input) do
    input |> parse_frequencies |> Enum.sum
  end

  def star_02(input) do
    frequencies = input |> parse_frequencies

    Stream.cycle(frequencies)
    |> Stream.scan(&(&1 + &2))
    |> first_repeat
  end

  def first_repeat(stream) do
    Enum.reduce_while(stream, MapSet.new, fn x, seen ->
      if x in seen do
        {:halt, x}
      else
        {:cont, MapSet.put(seen, x)}
      end
    end)
  end

  def parse_frequencies(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(fn x -> String.to_integer(x) end)
  end
end

input = File.read!("../inputs/day_01.txt")
592 = AdventOfCode.Day_01.star_01(input)
241 = AdventOfCode.Day_01.star_02(input)
